### ISSUE TITLE
Industrial Apiary Production Upgrades don't work as intended (have almost no effect) - RFC

### ISSUE THREAD

**Issue by hacatu** [GTNewHorizons/GT-New-Horizons-Modpack#12651]
### Your GTNH Discord Username

hacatu

### Your Pack Version

2.2.8-2.2.10

### Your Server

SP

### Java Version

Java 8

### Type of Server

None

### Your Expectation

Production upgrades should increase production significantly

### The Reality

The nerfed bee production formula in 2.2.8 is

`p = min((1 + t/6)*(1 + s)*2*sqrt(c) + (m/4)^cbrt(c) - 3, 100)`

where
* `t` is 1 for non-gt bee housings and 8 for the iapiary
* `s` is the speed of the bee (2.0 is best)
* `c` is the base chance of the product in question, as a percentage out of 100
* `m` is the production modifier for the bee housing
* `p` is the effective chance of the product in question, as a percentage out of 100

As you can see, if `m<=4`, the contribution of the second term is negligible (under 1% absolute/7.1% relative).

The iapiary calculates its production multiplier `m` as

`m = 0.25*u*b`

where
* `u` is the number of production upgrades installed
* `b` is the beekeeping mode production modifier (1 under default gtnh config)

See [iapiary m calculation](https://github.com/GTNewHorizons/GT5-Unofficial/blob/63ea934b93ae3d6b150fa067782d23d37cada453/src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_IndustrialApiary.java#L905),
[iapiary upgrade logic](https://github.com/GTNewHorizons/GT5-Unofficial/blob/63ea934b93ae3d6b150fa067782d23d37cada453/src/main/java/gregtech/api/util/GT_ApiaryUpgrade.java#L29).

In other words, the iapiary has a maximal production multiplier of 2, which corresponds to a maximal increase of 0.5% absolute/3.6% relative.

This doesn't seem intended.

### Your Proposal

The production upgrades should have a larger effect.

The tooltip says they are 1.2x.

I would probably have m start at 1 or 4, rather than 0, and then increase by ~1.892x or 1.2x if it starts at 4, or 1.4x - ~ 1.4142x if it starts at 1.  (The rough looking numbers basically send m to 16x, which is the cutoff above which bees normally can become ignoble, but this doesn't happen in the apiary so I'm just using it as a benchmark).

I can submit a PR to fix this myself, I just wanted to clarify exactly what the intended behavior is supposed to be.

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the bug still exists will prompt us to investigate and reopen it once we confirm your report.
- [X] I can reproduce this problem consistently by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.
- [X] I have asked other people and they confirm they also have this problem by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.

**Comment by chochem** [comment]
yea the new formula is pretty weird, not gonna lie. I told Runa as much at the time. Separating the production stat to an additive term like that makes no sense.

**Comment by Runakai1** [comment]
Intended behaviour

**Comment by boubou19** [comment]
i don't like this

**Comment by hacatu** [comment]
Also to clarify, a value of 16 for m would actually correspond to a 1.7x multiplier at most.  `p(c, 16)/p(c, 1)` increases in `c` until `p(c, 16)` hits the cap of 100, so we can find the max actual multiplier for any m by finding the `c` where `p(c, m) = 100` (via Newton's method or whatever) and then plugging that into `p(c, m)/p(c, 1)`.  For 16, it's 1.7x at c=19

**Comment by schneid3306** [comment]
So, not sure if this is the right ticket, but was told on discord there was an open ticket discussing the bee formula, and this is the only one I can find. I am on 2.2.10, and have had two bees running for about 6 hours. One is in a Magic Apiary (from what I have read 0.9x production) and one is in a regular Apiary (0.1x production) and the outputs are near identical for combs with a 15% chance (within 10 combs of each other). There were no boosters or frames impacting either Apiary. This doesn't seem correct, but I'm not sure if it is intended. In 2.2.3 I was getting many, many more combs from Magic Apiaries when I was in creative testing. I think the Magic Apiary was a nice middle ground between Alveary and Apiary with frames. I'd love for it to be worthwhile investing in, but in 2.2.10 it is not. I was planning on running a decent subset of bees off of the Magic Apiaries. Just my $0.02. Thanks.

**Comment by chochem** [comment]
new formula proposed by Runa is:
> 3.6093 x^(0.75)+ln(1+s^(4)+2 p^(1.25 l)) (1+p) s

I still think thats terrible as its split the production into 2 completely separate terms. None of the production modifiers actually modify the base chance in any way!

**Comment by chochem** [comment]
a good formula should start multiplacative, like
`p = x^c1 * (1+p)^c2 * s^c3 * c4`
and then you can play around with c1, c2, c3, c4 or replace something with a log to enact your nerf

**Comment by hacatu** [comment]
Agree with chochem.  I suggested

`(α(xpsl) + β)^γ` with `γ<1` or `αx(p^β)(1+(s-1)^γ)l` with `β, γ < 1` to begin with (again with both `x` and the output in percent)

a long time ago.  But in addition to having the production modifier and speed ... modify the chance based term instead of being in a separate term, I think the tier of WA upgrade `l` should not factor into the production formula, because it's already an exponential increase due to making the iapiary work exponentially faster.  Making it also boost production probability just seems redundant, although I understand Runakai's reasoning that it makes production upgrades more effective at higher tiers.  The speed boost already kind of does that though

**Comment by Runakai1** [comment]
> 3.6093 x^(0.75)+ln(1+s^(4)+2 p^(1.25 l)) (1+p) s

If anyone manages to make a multiplicative formula that
1) Allows for proper scaling. I dont like having a range of 4% comb chance until you hit the 100% cap
2) Allows for machine tier to be the major scaling factor
3) Properly manages to nerf steam to lategame in a reasonable way (Only things I encountered so far are earlygame too good/endgame too bad, both being too god, both being too bad)

**Comment by hacatu** [comment]
Edit: `l` aka `t` is machine tier, not number of WA upgrades, oops

**Comment by boubou19** [comment]
wtf am i looking at? can't we just define what's the defined production by tier/upgrade then use Lagrange's interpolation? that's what was done for the Tree Growth Simulator and it worked well, and the polynomial equation is dead simple.

**Comment by hacatu** [comment]
You mean [this](
https://github.com/GTNewHorizons/GTplusplus/blob/4b01cbd50843039fc0427f9de682fabfb0045da8/src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMetaTileEntityTreeFarm.java#L164)?  Yeah, a polynomial equation could work, although since dependence on the base production chance is supposed to be sublinear, polynomial interpolation isn't the best approach imo.  It would work, but the formula would be more obtuse

**Comment by chochem** [comment]
> Agree with chochem. I suggested
> 
> `(α(xpsl) + β)^γ` with `γ<1` or `αx(p^β)(1+(s-1)^γ)l` with `β, γ < 1` to begin with (again with both `x` and the output in percent)
> 
> a long time ago. But in addition to having the production modifier and speed ... modify the chance based term instead of being in a separate term, I think the tier of WA upgrade `l` should not factor into the production formula, because it's already an exponential increase due to making the iapiary work exponentially faster. Making it also boost production probability just seems redundant, although I understand Runakai's reasoning that it makes production upgrades more effective at higher tiers. The speed boost already kind of does that though

yes that sounds pretty good to me. its simple and effective. should be better to fit this kind of sublinear functions than a polynomial.

**Comment by kuba6000** [comment]
Bees production formula has been reworked:
https://github.com/GTNewHorizons/ForestryMC/pull/45
https://github.com/GTNewHorizons/GT5-Unofficial/pull/1820
